consul/state/prepared_query.go

```diff
@@ -0,0 +1,251 @@
+		return fmt.Errorf("invalid service %#v", query.Service.Service)
+	}
+
+	// Insert the query.
+	if err := tx.Insert("prepared-queries", query); err != nil {
+		return fmt.Errorf("failed inserting prepared query: %s", err)
+	}
+	if err := tx.Insert("index", &IndexEntry{"prepared-queries", idx}); err != nil {
+		return fmt.Errorf("failed updating index: %s", err)
+	}
+
+	tx.Defer(func() { s.tableWatches["prepared-queries"].Notify() })
+	return nil
+}
+
+// PreparedQueryDelete deletes the given query by ID.
+func (s *StateStore) PreparedQueryDelete(idx uint64, queryID string) error {
+	tx := s.db.Txn(true)
+	defer tx.Abort()
+
+	watches := NewDumbWatchManager(s.tableWatches)
+	if err := s.preparedQueryDeleteTxn(tx, idx, watches, queryID); err != nil {
+		return fmt.Errorf("failed prepared query delete: %s", err)
+	}
+
+	tx.Defer(func() { watches.Notify() })
+	tx.Commit()
+	return nil
+}
+
+// preparedQueryDeleteTxn is the inner method used to delete a prepared query
+// with the proper indexes into the state store.
+func (s *StateStore) preparedQueryDeleteTxn(tx *memdb.Txn, idx uint64, watches *DumbWatchManager,
+	queryID string) error {
+	// Pull the query.
+	query, err := tx.First("prepared-queries", "id", queryID)
+	if err != nil {
+		return fmt.Errorf("failed prepared query lookup: %s", err)
+	}
+	if query == nil {
+		return nil
+	}
+
+	// Delete the query and update the index.
+	if err := tx.Delete("prepared-queries", query); err != nil {
+		return fmt.Errorf("failed prepared query delete: %s", err)
+	}
+	if err := tx.Insert("index", &IndexEntry{"prepared-queries", idx}); err != nil {
+		return fmt.Errorf("failed updating index: %s", err)
+	}
+
+	watches.Arm("prepared-queries")
+	return nil
+}
+
+// PreparedQueryGet returns the given prepared query by ID.
+func (s *StateStore) PreparedQueryGet(queryID string) (uint64, *structs.PreparedQuery, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	// Get the table index.
+	idx := maxIndexTxn(tx, s.getWatchTables("PreparedQueryGet")...)
+
+	// Look up the query by its ID.
+	query, err := tx.First("prepared-queries", "id", queryID)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed prepared query lookup: %s", err)
+	}
+	if query != nil {
+		return idx, query.(*structs.PreparedQuery), nil
+	}
+	return idx, nil, nil
+}
+
+// PreparedQueryLookup returns the given prepared query by looking up an ID or
+// Name.
+func (s *StateStore) PreparedQueryLookup(queryIDOrName string) (uint64, *structs.PreparedQuery, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	// Get the table index.
+	idx := maxIndexTxn(tx, s.getWatchTables("PreparedQueryLookup")...)
+
+	// Explicitly ban an empty query. This will never match an ID and the
+	// schema is set up so it will never match a query with an empty name,
+	// but we check it here to be explicit about it (we'd never want to
+	// return the results from the first query w/o a name).
+	if queryIDOrName == "" {
+		return idx, nil, ErrMissingQueryID
```

Comment:
Should the returned index be 0 here? I think we use 0 in pretty much all error cases unless I'm missing something.